### PR TITLE
feat: group membership requests

### LINKED ISSUE #2241: Is there a way to approve/reject a membership group request ? Evaluation failed: H
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I'm trying to automatically accept/reject members who want to join a group. Is there a way to do that ? I didn't find any function to handle membership group requests

I tried to listen on 'group_update' (I correctly receive the notification) and then use addParticipants method from the correct group chat instance but I got 'Evaluation failed: H' error.

### Expected behavior

The user should be added in the group

### Steps to Reproduce the Bug or Issue

```
client.on('group_update', async (notification) => {
  console.log(notification);

  const chats = await client.getChats();
  const chat = chats.find((chat) => 'group_name' === chat.name) as GroupChat;

  // @ts-ignore
  const toto = await chat.addParticipants([notification.id.participant]);

  console.log(toto);
});
```

### Relevant Code

_No response_

### Browser Type

Chromium

### WhatsApp Account Type

WhatsApp Business

### Does your WhatsApp account have multidevice enabled?

No, I am not using Multi Device

### Environment

OS: mac
whatsapp-web.js: 1.20.0
WhatsApp Web version: 2.2322.15
Node.js: v16.17.0

### Additional context

_No response_

### LINKED ISSUE #2379: Feature request: methods to get, approve and reject requests to join a group
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

I have a group (with disappearing messages enabled) where users need admin approval before they can join. Unfortunately, I can't handle well all use cases with my current workaround.

### Describe the solution you'd like.

I need dedicated methods to get, approve and reject those requests

### Describe an alternate solution.

So far I have found 2 solutions to achieve this. The first solution is to listen for the 'group_update' event and check the notification type:

```
if (notification.type === 'created_membership_requests' || notification.type === ''membership_approval_request') {
// My code
}
```

I also need to achieve this outside of an event, the second solution for this case is to fetch the group messages and check the message subtype:

```
const joinRequestMessages = orderBy((await chat.fetchMessages({
  // @ts-ignore
  limit: 'Infinity',
})), ['timestamp'], ['desc']).filter(
  // @ts-ignore
  (message) => (message._data.subtype === 'created_membership_requests' || message._data.subtype === 'membership_approval_request')),
);

for (const joinRequestMessage of joinRequestMessages) {
// My code
}
```

The first problem with those solutions is that as you have seen it is a bit tricky to get messages and if my group has enabled disappearing messages, **once the messages are gone I can no longer fetch join requests.**

The second problem is that I cannot reject a request and to approve one I have to use the addParticipant method which **is not the correct way to approve a join request.**

Baileys library handle this with 2 methods: groupRequestParticipantsList and groupRequestParticipantsUpdate

Thank you !

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by alechkos**
# Table of Contents <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#toc" id="toc"/>

### - [How Has the PR Been Tested (latest test on 25.09.2023)](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#tests)
- [**The environment the PR was tested on**](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#env)
- [**I have an issue while testing this PR**](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#pr-testing-issue)

### - [Types of Changes](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#types-of-changes)

---

## Description [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#description" id="description"/>
The PR adds functionality to get, approve, reject and listen to group membership requests.
Things that were added:
1. An event to listen to new group membership requests
2. Three new methods both for `Client` and `GroupChat` classes (to execute on `GroupChat` object and without it):
    - `getGroupMembershipRequests` method that returns membership requests for specific group
    - `approveGroupMembershipRequests` method to approve requests from specific user, several users or all the existing membership requests
    - `rejectGroupMembershipRequests` the same as previous but to reject requests

More about `approveGroupMembershipRequests` and `rejectGroupMembershipRequests`:
- Both methods take:
    - group ID (if you are calling the method from `client` instance)
    - optional parameter `options`
- Options can take:
    - `requesterIds`: a single user ID as a string or an array of user IDs to approve/reject requests from
    - `sleep`: a sleep value
- The sleep value is necessary to avoid the risk of your accounts being banned. According to the logic of how these methods work in the WA, the operation is performed on only one user at a time, and we iterate over each user ID to perform the operation. If you pass more than one user ID, the method will, by default, choose a random sleep value between 250 to 500 milliseconds to wait before operation with next user. However, you have the flexibility to change this sleep value as needed
- If you want to approve or reject all existing membership requests for a specific group without specifying user IDs individually, you can simply call these methods without passing any user IDs
- An empty array will be returned in these cases:
    1. You wanted to approve/reject all the membership requests but there are no any
    2. A `SmaxParsingFailure` error was occupied

## Related Issues <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#related-issues" id="related-issues"/>
PR closes #2379 closes #2241

### Methods:
```js
client.on('message', async (msg) => {
    if (msg.body === '!approverequest') {
        /**
         * Presented an example for membership request approvals, the same examples are for the requesrejections.
         * To approve the membership request from a specific user:
         */
        await client.approveGroupMembershipRequests(msg.from, { requesterIds: 'number@c.us' });
        /** The same for execution on group object (no need to provide the group ID): */
        const group = await msg.getChat();
        await group.approveGroupMembershipRequests({ requesterIds: 'number@c.us' });
        /** To approve several membership requests: */
        const approval = await client.approveGroupMembershipRequests(msg.from, {
            requesterIds: ['number1@c.us', 'number2@c.us']
        });
        /**
         * The example of the {@link approval} output:
         * [
         *   {
         *     requesterId: 'number1@c.us',
         *     message: 'Rejected successfully'
         *   },
         *   {
         *     requesterId: 'number2@c.us',
         *     error: 404,
         *     message: 'ParticipantRequestNotFoundError'
         *   }
         * ]
         *
         */
        console.log(approval);
        /** To approve all the existing membership requests (simply don't provide any user IDs): */
        await client.approveGroupMembershipRequests(msg.from);
        /** To change the sleep value to 300 ms: */
        await client.approveGroupMembershipRequests(msg.from, {
            requesterIds: ['number1@c.us', 'number2@c.us'],
            sleep: 300
        });
        /** To change the sleep value to random value between 100 and 300 ms: */
        await client.approveGroupMembershipRequests(msg.from, {
            requesterIds: ['number1@c.us', 'number2@c.us'],
            sleep: [100, 300]
        });
        /** To explicitly disable the sleep: */
        await client.approveGroupMembershipRequests(msg.from, {
            requesterIds: ['number1@c.us', 'number2@c.us'],
            sleep: null
        });
    }
});
```

### Event:
```js
client.on('group_membership_request', async (notification) => {
    /**
     * The example of the {@link notification} output:
     * {
     *     id: {
     *         fromMe: false,
     *         remote: 'groupId@g.us',
     *         id: '123123123132132132',
     *         participant: 'number@c.us',
     *         _serialized: 'false_groupId@g.us_123123123132132132_number@c.us'
     *     },
     *     body: '',
     *     type: 'created_membership_requests',
     *     timestamp: 1694456538,
     *     chatId: 'groupId@g.us',
     *     author: 'number@c.us',
     *     recipientIds: []
     * }
     *
     */
    console.log(notification);
    /** You can approve or reject the newly appeared membership request: */
    await client.approveGroupMembershipRequests(notification.chatId, notification.author);
    await client.rejectGroupMembershipRequests(notification.chatId, notification.author);
});
```

## To test this PR by yourself you can run one of the following commands: <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#pr-testing" id="pr-testing"/>

- **NPM**
```powershell
npm install github:alechkos/whatsapp-web.js#grp-memb-req
```
- **YARN**
```powershell
yarn add github:alechkos/whatsapp-web.js#grp-memb-req
```

## How Has This Been Tested (latest test on 25.09.2023) [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#tests" id="tests"/>
The functionality has been tested with the code provided in usage example.

#### Types of accounts:
- Personal
- Buisness

#### Environment:
- Android 10
- Windows 10:
    - WWebJS **v1.22.2-alpha.1**
    - WWeb **v2.2333.11**, **v2.2340.11**
    - Puppeteer **v18.2.1**
    - Node **v18.17.1**
    - Chrome **v117.0.5938.92**

<details><summary>

### If you encounter any issues while testing this PR, please provide in a comment (click to expand): <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#pr-testing-issue" id="pr-testing-issue"/>
</summary>

> [!IMPORTANT]
> **You have to [reapply](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#pr-testing) the PR each time it is changed (new commits were pushed since your last application)**

1. The code you've used without any sensitive information (use [syntax highlighting](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting) for more readability)
4. The library version
5. The WWeb version: `console.log(await client.getWWebVersion());`
6. The browser (Chrome/Chromium)
7. The error you got
</details>

## Types of Changes [_[Back to TOC]_](https://github.com/pedroslopez/whatsapp-web.js/pull/2516#toc) <a href="https://github.com/pedroslopez/whatsapp-web.js/pull/2516#types-of-changes" id="types-of-changes"/>
- [ ] Dependency change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix that would cause existing functionality to change)

## Checklist
- [X] My code follows the code style of this project
- [X] I have updated the usage example accordingly (example.js)
- [X] I have updated the documentation accordingly (index.d.ts)

### CODE REVIEW

**Comment by Bauguelson** [comment]
Not working for me: 
Error: Evaluation failed: TypeError: Cannot read properties of undefined (reading 'widToUserJid')

**Comment by alechkos** [comment]
@Bauguelson 
> Error: Evaluation failed: TypeError: Cannot read properties of undefined (reading 'widToUserJid')

fixed

**Review by shirser121 — APPROVED** [review]
Tested and work perfectly 
Good job

**Comment by shirser121** [comment]
@aliyss @tuyuribr @PurpShell

**Review by tuyuribr — APPROVED** [review]
Good work

File: index.d.ts [hunk 1]
```diff
@@ -1273,17 +1324,35 @@ declare namespace WAWebJS {
...
+         * @returns {Promise<Array<MembershipRequestActionResult>>} Returns an array of requester IDs whose membership requests were approved and an error for each requester, if any occurred during the operation. If there are no requests, an empty array will be returned
+         */
+        approveGroupMembershipRequests: (options: MembershipRequestActionOptions) => Promise<Array<MembershipRequestActionResult>>;
+        /**
+         * Rejects membership requests if any
+         * @param {MembershipRequestActionOptions} options Options for performing a membership request action
+         * @returns {Promise<Array<MembershipRequestActionResult>>} Returns an array of requester IDs whose membership requests were rejected and an error for each requester, if any occurred during the operation. If there are no requests, an empty array will be returned
+         */
+        rejectGroupMembershipRequests: (options: MembershipRequestActionOptions
+        ) => Promise<Array<MembershipRequestActionResult>>;
```
[the remark this answers is not in the document]
Comment (alechkos) [reply index.d.ts:1355]: @tuyuribr Done

**Comment by zvictor** [comment]
@alechkos what a great PR description!
I also noticed you have pushed quite a few things recently to this project. Thank you so much! ❤️

**Pull request merged**